catenets/models/torch/pseudo_outcome_nets.py

```diff
@@ -0,0 +1,623 @@
+            weight_decay=self.weight_decay,
+            lr=self.lr,
+            n_iter=self.n_iter,
+            batch_size=self.batch_size,
+            val_split_prop=self.val_split_prop,
+            n_iter_print=self.n_iter_print,
+            seed=self.seed,
+            nonlin=self.nonlin,
+            patience=self.patience,
+            n_iter_min=self.n_iter_min,
+        ).to(DEVICE)
+
+    def _generate_po_estimator(self, name: str = "po_estimator") -> nn.Module:
+        if self._po_template is not None:
+            return copy.deepcopy(self._po_template)
+
+        return BasicNet(
+            name,
+            self.n_unit_in,
+            binary_y=self.binary_y,
+            n_layers_out=self.n_layers_out_t,
+            n_units_out=self.n_units_out_t,
+            weight_decay=self.weight_decay_t,
+            lr=self.lr_t,
+            n_iter=self.n_iter,
+            batch_size=self.batch_size,
+            val_split_prop=self.val_split_prop,
+            n_iter_print=self.n_iter_print,
+            seed=self.seed,
+            nonlin=self.nonlin,
+            patience=self.patience,
+            n_iter_min=self.n_iter_min,
+        ).to(DEVICE)
+
+    def _generate_propensity_estimator(
+        self, name: str = "propensity_estimator"
+    ) -> nn.Module:
+        if self.weighting_strategy is None:
+            raise ValueError("Invalid weighting_strategy for PropensityNet")
+        return PropensityNet(
+            name,
+            self.n_unit_in,
+            2,  # number of treatments
+            self.weighting_strategy,
+            n_units_out_prop=self.n_units_out_prop,
+            n_layers_out_prop=self.n_layers_out_prop,
+            weight_decay=self.weight_decay,
+            lr=self.lr,
+            n_iter=self.n_iter,
+            batch_size=self.batch_size,
+            n_iter_print=self.n_iter_print,
+            seed=self.seed,
+            nonlin=self.nonlin,
+            val_split_prop=self.val_split_prop,
+        ).to(DEVICE)
+
+    def train(
+        self, X: torch.Tensor, y: torch.Tensor, w: torch.Tensor
+    ) -> "PseudoOutcomeLearner":
+        """
+        Train treatment effects nets.
+
+        Parameters
+        ----------
+        X: array-like of shape (n_samples, n_features)
+            Train-sample features
+        y: array-like of shape (n_samples,)
+            Train-sample labels
+        w: array-like of shape (n_samples,)
+            Train-sample treatments
+        """
+        X = torch.Tensor(X).to(DEVICE)
+        y = torch.Tensor(y).squeeze().to(DEVICE)
+        w = torch.Tensor(w).squeeze().to(DEVICE)
+
+        n = len(y)
+
+        # STEP 1: fit plug-in estimators via cross-fitting
+        mu_0_pred, mu_1_pred, p_pred = (
+            torch.zeros(n).to(DEVICE),
+            torch.zeros(n).to(DEVICE),
+            torch.zeros(n).to(DEVICE),
+        )
+
+        # create folds stratified by treatment assignment to ensure balance
+        splitter = StratifiedKFold(
```

Comment:
I think this might not work if n_folds=1? (Sklearn documentation says n_folds>=2). That's why I had an if-statement originally.